providers/common/ciphers/aes.c

```diff
@@ -340,50 +346,43 @@ static void *aes_dupctx(void *ctx)
     return ret;
 }
 
-static size_t key_length_256(void)
-{
-    return 256 / 8;
-}
-
-static size_t key_length_192(void)
-{
-    return 192 / 8;
-}
-
-static size_t key_length_128(void)
-{
-    return 128 / 8;
-}
-
-static size_t iv_length_16(void)
-{
-    return 16;
-}
-
-static size_t iv_length_0(void)
-{
-    return 0;
-}
-
-static size_t block_size_16(void)
-{
-    return 16;
-}
-
-static size_t block_size_1(void)
-{
-    return 1;
-}
-
 static int aes_ctx_get_params(void *vctx, OSSL_PARAM params[])
 {
     PROV_AES_KEY *ctx = (PROV_AES_KEY *)vctx;
     OSSL_PARAM *p;
 
     p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_PADDING);
-    if (p != NULL && !OSSL_PARAM_set_int(p, ctx->pad)) {
-        PROVerr(PROV_F_AES_CTX_GET_PARAMS, PROV_R_FAILED_TO_SET_PARAMETER);
-        return 0;
+    if (p != NULL) {
+        if (!OSSL_PARAM_set_int(p, ctx->pad)) {
+            PROVerr(PROV_F_AES_CTX_GET_PARAMS,
+                    PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+    }
+    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IV);
+    if (p != NULL) {
+        if (!OSSL_PARAM_set_octet_ptr(p, &ctx->iv, AES_BLOCK_SIZE)
```

Comment:
Should these have the same name?  I'm thinking of the `OSSL_ITEM param_types[]` arrays being queried.  I think that makes sense for things other than core and providers.

Could the PTR form always be used and have a copy in the ctrl wrapper?

Reply:
I think they should be seperate.
As an aside - I think where you declare this define you only have the one type associated with it..

Reply:
Why have separate names?  The IV is the IV is the IV, no matter how you choose to transfer it.

Reply:
I think it makes it clearer that the type is different.

Reply:
> Could the PTR form always be used and have a copy in the ctrl wrapper?

Personally, I'm not a fan of handing over a pointer to the live IV.  I'm pretty damn sure that some provider authors will refuse to support such a thing (I know for a fact that it will be an impossible thing if the provider resides in a protected environment).  However, we do have `EVP_CIPHER_CTX_iv`, which does exactly this, so as long as that function exists and is used (and it is...  it was a report about this function not working right that got me started on this PR).

Reply:
Fair point, it was a suggestion not mandatory.
The alternative could be a param type check but that complicates the code a bit (and I'm not that fussed).

We could have `EVP_CIPHER_CTX_iv` fail for algorithms that don't support this functionality without breaking the API.

Reply:
> We could have `EVP_CIPHER_CTX_iv` fail for algorithms that don't support this functionality without breaking the API.

That's what this line does:

```
    return ok != 0 ? v: NULL;
```

(`ok` may be -2, which is generally used for "unsupported", which I use in this case to signal that this isn't a provider implementation)